src/crimson/os/seastore/lba_manager/btree/btree_range_pin.h

```diff
@@ -44,41 +60,191 @@ inline std::ostream &operator<<(
+    : range(rhs.range), extent(extent) {}
+
+  bool has_ref() const {
+    return !!ref;
+  }
+
+  bool is_root() const {
+    return range.is_root();
+  }
+
+  void set_range(const lba_node_meta_t &nrange) {
+    range = nrange;
+  }
+  void set_extent(CachedExtent *nextent) {
+    assert(!extent);
+    extent = nextent;
+  }
+
+  friend bool operator<(
+    const btree_range_pin_t &lhs, const btree_range_pin_t &rhs) {
+    return get_tuple(lhs.range) < get_tuple(rhs.range);
+  }
+  friend bool operator>(
+    const btree_range_pin_t &lhs, const btree_range_pin_t &rhs) {
+    return get_tuple(lhs.range) > get_tuple(rhs.range);
+  }
+  friend bool operator==(
+    const btree_range_pin_t &lhs, const btree_range_pin_t &rhs) {
+    return get_tuple(lhs.range) == rhs.get_tuple(rhs.range);
+  }
+
+  struct meta_cmp_t {
+    bool operator()(
+      const btree_range_pin_t &lhs, const lba_node_meta_t &rhs) const {
+      return get_tuple(lhs.range) < get_tuple(rhs);
+    }
+    bool operator()(
+      const lba_node_meta_t &lhs, const btree_range_pin_t &rhs) const {
+      return get_tuple(lhs) < get_tuple(rhs.range);
+    }
+  };
+
+  friend std::ostream &operator<<(
+    std::ostream &lhs,
+    const btree_range_pin_t &rhs) {
+    return lhs << "btree_range_pin_t("
+	       << "begin=" << rhs.range.begin
+	       << ", end=" << rhs.range.end
+	       << ", depth=" << rhs.range.depth
+	       << ", extent=" << rhs.extent
+	       << ")";
+  }
+
+  friend class BtreeLBAPin;
+  ~btree_range_pin_t();
+};
+
+/**
+ * btree_pin_set_t
  *
- * References leaf node
+ * Contains a btree_range_pin_t for every clean or dirty LBANode
+ * or LogicalCachedExtent instance in cache at any point in time.
  *
- * TODO: does not at this time actually keep the relevant
- * leaf resident in memory.  This is actually a bit tricky
- * as we can mutate and therefore replace a leaf referenced
- * by other, uninvolved but cached extents.  Will need to
- * come up with some kind of pinning mechanism that handles
- * that well.
+ * For any LBANode, the contained btree_range_pin_t will hold
+ * a reference to that node pinning it in cache as long as that
+ * node has children in the set.  This invariant can be violated
+ * only by calling retire_extent and is repaired by calling
+ * check_parent syncronously after adding any new extents.
  */
+class btree_pin_set_t {
+  friend class btree_range_pin_t;
+  btree_range_pin_t::index_t pins;
+
+  decltype(pins)::iterator get_iter(btree_range_pin_t &pin) {
+    return decltype(pins)::s_iterator_to(pin);
+  }
+
+  /// Removes pin from set optionally checking whether parent has other children
+  void remove_pin(btree_range_pin_t &pin, bool check_parent);
+
+  /// Returns parent pin if exists
+  btree_range_pin_t *maybe_get_parent(const lba_node_meta_t &pin);
+
+  /// Returns earliest child pin if exist
+  btree_range_pin_t *maybe_get_first_child(const lba_node_meta_t &pin);
```

Comment:
```suggestion
  const btree_range_pin_t *maybe_get_first_child(const lba_node_meta_t &pin) const;
```

Reply:
Not const, the caller may modify them.

Reply:
but i cannot find a use case where the caller changes the returned child pin.

Reply:
Hmm, maybe_get_parent does get used to increment the ref count.  I guess can make maybe_get_first_child const.